eip-0023.md

```diff
@@ -0,0 +1,584 @@
+  // R4 the group element of the owner of the ballot token [GroupElement]
+  // R5 the creation height of the update box [Int]
+  // R6 the value voted for [Coll[Byte]]
+  // R7 the reward token id [Coll[Byte]]
+  // R8 the reward token amount [Long]
+
+  val updateNFT = fromBase64("YlFlVGhXbVpxNHQ3dyF6JUMqRi1KQE5jUmZValhuMnI=") // TODO replace with actual 
+
+  val minStorageRent = 10000000L  // TODO replace with actual
+  
+  val selfPubKey = SELF.R4[GroupElement].get
+  
+  val outIndex = getVar[Int](0).get
+  val output = OUTPUTS(outIndex)
+  
+  val isSimpleCopy = output.R4[GroupElement].isDefined                && // ballot boxes are transferable by setting different value here 
+                     output.propositionBytes == SELF.propositionBytes &&
+                     output.tokens == SELF.tokens                     && 
+                     output.value >= minStorageRent 
+  
+  val update = INPUTS.size > 1                           &&
+               INPUTS(1).tokens.size > 0                 &&
+               INPUTS(1).tokens(0)._1 == updateNFT       && // can only update when update box is the 2nd input
+               output.R4[GroupElement].get == selfPubKey && // public key is preserved
+               output.value >= SELF.value                && // value preserved or increased
+               ! (output.R5[Any].isDefined)                 // no more registers; prevents box from being reused as a valid vote 
+  
+  val owner = proveDlog(selfPubKey)
+  
+  // unlike in collection, here we don't require spender to be one of the ballot token holders
+  isSimpleCopy && (owner || update)
+}
+```
+
+### Update Contract
+
+```
+{ // This box (update box):
+  // Registers empty 
+  // 
+  // ballot boxes (Inputs)
+  // R4 the pub key of voter [GroupElement] (not used here)
+  // R5 the creation height of this box [Int]
+  // R6 the value voted for [Coll[Byte]] (hash of the new pool box script)
+  // R7 the reward token id in new box 
+  // R8 the number of reward tokens in new box 
+
+  val poolNFT = fromBase64("RytLYlBlU2hWbVlxM3Q2dzl6JEMmRilKQE1jUWZUalc=") // TODO replace with actual 
+
+  val ballotTokenId = fromBase64("P0QoRy1LYVBkU2dWa1lwM3M2djl5JEImRSlIQE1iUWU=") // TODO replace with actual 
+
+  val minVotes = 6 // TODO replace with actual
+  
+  val poolIn = INPUTS(0) // pool box is 1st input
+  val poolOut = OUTPUTS(0) // copy of pool box is the 1st output
+
+  val updateBoxOut = OUTPUTS(1) // copy of this box is the 2nd output
+
+  // compute the hash of the pool output box. This should be the value voted for
+  val poolOutHash = blake2b256(poolOut.propositionBytes)
+  val rewardTokenId = poolOut.tokens(1)._1
+  val rewardAmt = poolOut.tokens(1)._2
+  
+  val validPoolIn = poolIn.tokens(0)._1 == poolNFT
+  
+  val validPoolOut = poolIn.tokens(0) == poolOut.tokens(0)                && // NFT preserved
+                     poolIn.creationInfo._1 == poolOut.creationInfo._1    && // creation height preserved
```

Comment:
@scalahub After EIP-39 (creation height of the output cannot be less than max input box creation height) implemented in https://github.com/ergoplatform/ergo/pull/1865 this requirement (pool box creation height preservation) is no longer satisfiable. We need to come up with a new requirement. What was the reason for this requirement in the first place?

Reply:
@greenhat I have removed the creation height preserving condition. Please review. Note that I've not verified that the individual contracts are same as the one in the single file.